### PR TITLE
feat(SystemsTable): RHICOMPL-1657 - Pass filters and orderBy to API

### PR DESCRIPTION
**Pull request by bastilian**
This adds sorting to the SystemsTable. For now only for the name and operating system column, and only in the Wizard, but we can gradually add more columns when we implement the backend part for them.

https://user-images.githubusercontent.com/7757/122219630-e9150680-ceaf-11eb-9e7a-703117929705.mp4

## Secure Coding Practices Checklist GitHub Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist
- [x] Input Validation
- [x] Output Encoding
- [x] Authentication and Password Management
- [x] Session Management
- [x] Access Control
- [x] Cryptographic Practices
- [x] Error Handling and Logging
- [x] Data Protection
- [x] Communication Security
- [x] System Configuration
- [x] Database Security
- [x] File Management
- [x] Memory Management
- [x] General Coding Practices

### CODE REVIEW

File: src/SmartComponents/SystemsTable/hooks.js [hunk 1]
```diff
@@ -84,8 +85,11 @@ export const useFetchSystems = ({
 };
 
 export const useGetEntities = (fetchEntities, { selected }) => (
-    async (_ids, { page = 1, per_page: perPage }) => {
-        const fetchedEntities = await fetchEntities(perPage, page);
+    async (_ids, { page = 1, per_page: perPage, orderBy, orderDirection, filters }) => {
+        const fetchedEntities = await fetchEntities(perPage, page, {
+            filters,
```
Comment (bastilian) [inline src/SmartComponents/SystemsTable/hooks.js:90]: We would not need this, and it doesn't do anything yet, but while we are here.

File: src/SmartComponents/SystemsTable/hooks.js [hunk 2]
```diff
@@ -84,8 +85,11 @@ export const useFetchSystems = ({
...
+            orderBy: `${orderBy} ${orderDirection}`
```
Comment (bastilian) [inline src/SmartComponents/SystemsTable/hooks.js:91]: We can also split them into two variables.

Comment (akofink) [reply src/SmartComponents/SystemsTable/hooks.js]: Shouldn't this use the colon syntax, @skateman?

```js
            orderBy: `${orderBy}:${orderDirection}`
```

Comment (bastilian) [reply src/SmartComponents/SystemsTable/hooks.js]: It does. This is a comment on an old outdated commit. :) I tried deleting it but GitHub apparently didn't let me.

Comment (skateman) [reply src/SmartComponents/SystemsTable/hooks.js]: Dfeinitely a colon

File: src/SmartComponents/SystemsTable/hooks.js [hunk 3]
```diff
@@ -83,20 +84,38 @@ export const useFetchSystems = ({
...
+    const appendDirection = (attributes, direction) => (
+        attributes.map((attribute) => `${attribute}:${direction}`)
+    );
+
+    const findColumnByKey = (key) => (
+        (columns || []).find((column) => column.key === key)
+    );
+
+    return async (_ids, { page = 1, per_page: perPage, orderBy, orderDirection, filters }) => {
+        const columnKey = orderBy.split(' ')[0];
```
Comment (vkrizan) [inline src/SmartComponents/SystemsTable/hooks.js:97]: Can you please elaborate on the split by a space and taking the first element?

File: src/SmartComponents/CreatePolicy/EditPolicySystems.js [hunk 4]
```diff
@@ -67,10 +67,17 @@ export const EditPolicySystems = ({ change, osMajorVersion, selectedSystems }) =
...
-                        columns={[
-                            Columns.Name,
-                            Columns.OperatingSystem
-                        ]}
+                        columns={[{
+                            ...Columns.Name,
+                            props: {
+                                width: 40
+                            },
+                            sortBy: ['name']
```
Comment (vkrizan) [inline src/SmartComponents/CreatePolicy/EditPolicySystems.js:75]: Can you also please add this column configuration to the EditPolicy modal? It is possible to click there to sort, but no sorting is happening.

File: src/SmartComponents/SystemsTable/hooks.js [hunk 5]
```diff
@@ -83,20 +84,38 @@ export const useFetchSystems = ({
...
+        const sortableColumn = findColumnByKey(columnKey);
+        const sortBy = sortableColumn && sortableColumn.sortBy ?
+            appendDirection(sortableColumn.sortBy, orderDirection) : undefined;
+
+        const fetchedEntities = await fetchEntities(perPage, page, {
+            filters,
+            sortBy
```
Comment (vkrizan) [inline src/SmartComponents/SystemsTable/hooks.js:104]: Once an order is changed, the whole table goes to preload state and hides the table header and page information for a brief moment. It jumps around. I'd expect only the table contents to be preloaded keeping the table header and page stable.

(line unrelated)

File: src/SmartComponents/CreatePolicy/EditPolicySystems.js [hunk 6]
```diff
@@ -67,10 +67,17 @@ export const EditPolicySystems = ({ change, osMajorVersion, selectedSystems }) =
...
+                        }, {
+                            ...Columns.OperatingSystem,
+                            props: {},
+                            sortBy: ['osMajorVersion', 'osMinorVersion']
```
Comment (vkrizan) [inline src/SmartComponents/CreatePolicy/EditPolicySystems.js:79]: Observation: DRY.The sorting array for common coulumns would likely be repeated wherever the column is used.

**Review by vkrizan — COMMENTED** [review]
So far so good.

The sorting works only in CreatePolicy, in EditPolicy it only gives the ability, and other system lists on Systems page and Policy's systems do not have any sorting enabled at all.

File: src/SmartComponents/SystemsTable/hooks.js (hunk 3 above)
Comment (bastilian) [reply src/SmartComponents/SystemsTable/hooks.js]: Remnant from earlier development.

File: src/SmartComponents/SystemsTable/hooks.js (hunk 5 above)
Comment (bastilian) [reply src/SmartComponents/SystemsTable/hooks.js:103]: That is the default behaviour of the Inventory table. We can bring it up with UX and the platform team.

https://user-images.githubusercontent.com/7757/122388071-cf3af880-cf6f-11eb-93f1-555c78991bec.mov

**Comment by bastilian** [comment]
@vkrizan Thank you for taking a look and reviewing this! RHICOMPL-1657 is only about adding sorting for the SystemsTable in the Wizard, that it was showing up on the EditPolicy was a bug with the columns being cached in the Inventory store. We will be adding sorting to the other SystemsTables and for all columns too, but in time. For the EditPolicy we also would want to be able and not only filter but also sort selected systems first.

**Review by akofink — APPROVED** [review]
Excellent! Works as expected. Any reason we shouldn't enable this on EditPolicy, PolicyDetails, and the Systems page as well?

**Comment by bastilian** [comment]
@akofink Thanks for taking a look as well! We can consider adding it to the Edit Policy, maybe the PolicyDetails, since they have the same columns and all columns in the table would be sortable. On the other tables I left them out to avoid a mix of sortable and none-sortable columns, as we are for now only able to sort by name and OS versions, while columns using other attributes would still be static.

**Comment by akofink** [comment]
Ah, makes sense! Yes, it'd be confusing if only some columns were sortable.

**Pull request merged**